synapse/logging/context.py

```diff
@@ -720,7 +733,30 @@ def g(*args, **kwargs):
     return g
 
 
-def run_in_background(f, *args, **kwargs) -> defer.Deferred:
+@overload
+def run_in_background(  # type: ignore[misc]
+    f: Callable[..., Awaitable[R]], *args: Any, **kwargs: Any
+) -> "defer.Deferred[R]":
+    # The `type: ignore[misc]` above suppresses
+    # "Overloaded function signatures 1 and 2 overlap with incompatible return types"
+    ...
+
+
+@overload
+def run_in_background(
+    f: Callable[..., R], *args: Any, **kwargs: Any
+) -> "defer.Deferred[R]":
+    ...
+
+
+def run_in_background(
+    f: Union[
+        Callable[..., R],
+        Callable[..., Awaitable[R]],
+    ],
```

Comment:
+764 seems to suggest that `f` could return a `defer.Deferred[R]` too.

(Is Deferred[T] considered compatible with Awaitable[T]?)

Reply:
It is: https://github.com/twisted/twisted/blob/trunk/src/twisted/internet/defer.py#L347

The `T` in `Deferred[T]` is contravariant for some reason. If it weren't contravariant, we wouldn't need the `type: ignore`s for the overloads, for reasons I haven't figured out yet.

Reply:
Oh! Well that makes life much easier!

> The `T` in `Deferred[T]` is contravariant for some reason.

Maybe because the `T` gets passed to a callback, and `Callable[T, ...]` is contravariant in `T`?

Reply:
That makes sense!